core/state/trie_prefetcher.go

```diff
@@ -123,47 +120,82 @@ func (p *TriePrefetcher) Loop() {
 				}
 				paused = false
 			} else {
+				// Prefetcher is being paused, abort all unfulfilled requests
+				p.taskLock.Lock()
+				skipped += int64(len(p.accountTasks))
+				for storageTasks := range p.storageTasks {
+					skipped += int64(len(storageTasks))
+				}
+				p.accountTasks = nil
+				p.storageTasks = make(map[common.Hash][]common.Hash)
+				p.taskLock.Unlock()
+
 				// Update metrics at new block events
 				triePrefetchFetchMeter.Mark(fetched)
 				triePrefetchSkipMeter.Mark(skipped)
 				fetched, skipped = 0, 0
+
 				// Make the tries accessible
 				p.accountTrie = accountTrie
 				p.storageTries = storageTries
 				p.accountTrieRoot = accountTrieRoot
+
 				if cmd.root != (common.Hash{}) {
 					log.Error("Trie prefetcher paused with non-empty root")
 				}
 				paused = true
 			}
-			p.deliveryCh <- struct{}{}
-		case req := <-p.requestCh:
+			close(cmd.done)
+
+		case <-p.reqCh:
 			if paused {
+				log.Error("Prefetch request arrived whilst paused")
 				continue
 			}
-			if sRoot := req.storageRoot; sRoot != nil {
-				// Storage slots to fetch
-				var (
-					storageTrie Trie
-					err         error
-				)
-				if storageTrie = storageTries[*sRoot]; storageTrie == nil {
-					if storageTrie, err = p.db.OpenTrie(*sRoot); err != nil {
-						log.Warn("trie prefetcher failed opening storage trie", "root", *sRoot, "err", err)
-						skipped += int64(len(req.slots))
+			// Retrieve all the tasks queued up and reset the sets for new insetions
+			p.taskLock.Lock()
+			accountTasks, storageTasks := p.accountTasks, p.storageTasks
+			if len(accountTasks) > 0 {
+				p.accountTasks = nil
+			}
+			if len(storageTasks) > 0 {
+				p.storageTasks = make(map[common.Hash][]common.Hash)
+			}
+			p.taskLock.Unlock()
+
+			// Keep prefetching the data until an interruption is triggered
+			for _, addr := range accountTasks {
+				if atomic.LoadUint32(&p.paused) == 1 {
+					break
+				}
+				accountTrie.TryGet(addr[:])
+				fetched++
+			}
+			for root, slots := range storageTasks {
+				if _, ok := storageTries[root]; !ok {
+					storageTrie, err := p.db.OpenTrie(root)
+					if err != nil {
+						log.Warn("Trie prefetcher failed opening storage trie", "root", root, "err", err)
+						skipped += int64(len(slots))
 						continue
 					}
-					storageTries[*sRoot] = storageTrie
+					storageTries[root] = storageTrie
 				}
-				for _, key := range req.slots {
-					storageTrie.TryGet(key[:])
+				storageTrie := storageTries[root]
+				for _, hash := range slots {
+					if atomic.LoadUint32(&p.paused) == 1 {
+						break
```

Comment:
This will only break out of the innermost loop.. is that the intent?

Reply:
Shit, no

Reply:
Fixed.